pkg/logs/tag/provider.go

```diff
@@ -28,32 +30,94 @@ var NoopProvider Provider = &noopProvider{}
 
 // provider caches a list of up-to-date tags for a given entity polling periodically the tagger.
 type provider struct {
-	entityID string
-	tags     []string
-	done     chan struct{}
-	mu       sync.Mutex
+	entityID             string
+	tags                 []string
+	done                 chan struct{}
+	forceRefresh         bool
+	waitOnFirstCall      bool
+	forceRefreshDuration time.Duration
+	taggerWarmupDuration time.Duration
+	mu                   sync.Mutex
 }
 
 // NewProvider returns a new Provider.
 func NewProvider(entityID string) Provider {
-	return &provider{
-		entityID: entityID,
-		tags:     []string{},
-		done:     make(chan struct{}),
+	p := &provider{
+		entityID:        entityID,
+		tags:            []string{},
+		done:            make(chan struct{}),
+		forceRefresh:    config.Datadog.GetBool("logs_config.k8s_wait_for_tags"),
+		waitOnFirstCall: config.Datadog.GetBool("logs_config.k8s_wait_for_tags"),
 	}
+	if p.forceRefresh {
+		p.forceRefreshDuration = config.Datadog.GetDuration("logs_config.force_tagger_call_duration") * time.Second
+		p.taggerWarmupDuration = config.Datadog.GetDuration("logs_config.tagger_warmup_duration") * time.Second
+	}
+	return p
 }
 
 // GetTags returns the list of up-to-date tags.
+// if logs_config.k8s_wait_for_tags enabled:
+//   - GetTags waits on its first call before updating the local cache for 2 seconds (default)
+//   - GetTags forces updating the local cache for 5 seconds (default)
 func (p *provider) GetTags() []string {
+	if p.getWaitOnFirstCall() {
+		// Warmup duration
+		// Make sure the tagger collects all the service tags
+		<-time.After(p.taggerWarmupDuration)
+		p.setWaitOnFirstCall(false)
+	}
+	if p.getForceRefresh() {
+		p.updateTags()
+	}
 	p.mu.Lock()
 	defer p.mu.Unlock()
 	return p.tags
 }
 
+// setForceRefresh udpates the forceRefresh attribute
+func (p *provider) setForceRefresh(b bool) {
```

Comment:
Why do we need all those setters and getters ?

Reply:
- `forceRefresh` is accessed by multiple functions in different goroutines, I implemented the getters and setters to ensure protecting it by the mutex

- `waitOnFirstCall` is no longer used and will be removed with its setter and getter (thanks to using `sync.Once`)